lib/grandpa/finalization.go

```diff
@@ -0,0 +1,493 @@
+			return
+		default:
+		}
+
+		err := fh.initiateRound()
+		if err != nil {
+			fh.observableErrs <- fmt.Errorf("initiating round: %w", err)
+			return
+		}
+
+		err = fh.waitServices()
+		if err != nil {
+			fh.observableErrs <- err
+			return
+		}
+	}
+}
+
+func (fh *finalizationHandler) stop() (err error) {
+	fh.servicesLock.Lock()
+	defer fh.servicesLock.Unlock()
+
+	stopWg := new(sync.WaitGroup)
+	stopWg.Add(2)
+
+	errs := [2]error{}
+	go func() {
+		defer stopWg.Done()
+		errs[0] = fh.finalizationEngine.Stop()
+	}()
+
+	go func() {
+		defer stopWg.Done()
+		errs[1] = fh.votingRound.Stop()
+	}()
+
+	stopWg.Wait()
+
+	if errs[0] != nil {
+		return errs[0]
+	}
+
+	if errs[1] != nil {
+		return errs[1]
+	}
+
+	return nil
+}
+
+func (fh *finalizationHandler) Stop() (err error) {
+	close(fh.stopCh)
+
+	err = fh.stop()
+	select {
+	case <-fh.handlerDone:
+	case <-time.After(fh.timeoutStop):
+		return errTimeoutWhileStoping
+	}
+
+	close(fh.observableErrs)
+	return err
+}
+
+// waitServices will start the services and wait until they complete or
+func (fh *finalizationHandler) waitServices() error {
+	fh.servicesLock.Lock()
+	fh.finalizationEngine, fh.votingRound = fh.newServices()
+	fh.servicesLock.Unlock()
+
+	finalizationEngineErr := make(chan error)
+	go func() {
+		defer close(finalizationEngineErr)
+		err := fh.finalizationEngine.Start()
+		if err != nil {
+			finalizationEngineErr <- err
+		}
+	}()
+
+	votingRoundErr := make(chan error)
+	go func() {
+		defer close(votingRoundErr)
+		err := fh.votingRound.Start()
+		if err != nil {
+			votingRoundErr <- err
+		}
+	}()
+
+	for {
+		select {
+		case <-fh.stopCh:
+			return nil
+
+		case err, ok := <-votingRoundErr:
+			if !ok {
+				votingRoundErr = nil
+			}
+
+			stopErr := fh.stop()
+			if stopErr != nil {
+				logger.Infof("stopping finalisation handler: %s", stopErr)
+			}
+			return err
+
+		case err, ok := <-finalizationEngineErr:
+			if !ok {
+				finalizationEngineErr = nil
+			}
+
+			stopErr := fh.stop()
+			if stopErr != nil {
+				logger.Infof("stopping finalisation handler: %s", stopErr)
+			}
+
+			return err
+
+		default:
+			finish := votingRoundErr == nil && finalizationEngineErr == nil
+			if finish {
+				return nil
+			}
+		}
+	}
+}
+
+var errTimeoutWhileStoping = errors.New("timeout while stopping")
+
+type handleVotingRound struct {
+	grandpaService       *Service
+	timeoutStop          time.Duration
+	finalizationEngineCh <-chan engineAction
+	stopCh               chan struct{}
+	engineDone           chan struct{}
+}
+
+func newHandleVotingRound(service *Service, finalizationEngineCh <-chan engineAction) *handleVotingRound {
+	return &handleVotingRound{
+		timeoutStop:          5 * time.Second,
+		grandpaService:       service,
+		stopCh:               make(chan struct{}),
+		engineDone:           make(chan struct{}),
+		finalizationEngineCh: finalizationEngineCh,
+	}
+}
+
+func (h *handleVotingRound) Stop() (err error) {
+	if h.stopCh == nil {
+		return nil
+	}
```

Comment:
it is not possible to check if a channel is closed by `_, ok := h.stopCh` we should use the arrow `<-` like: `_, ok := <-h.stopCh` but we cannot do that here since it will block until the `stopCh` is closed so we face the problem where in the first time we call the `Stop` method it will block forever, however setting it to `nil` once the channel is closed enable us to do the check without blocking.

Reply:
Actually you can use

```go
var ok bool
select {
case _, ok := <- h.stopCh
default:
}
```